packages/common/src/hooks/useHarmonyField.ts

```diff
@@ -0,0 +1,262 @@
+      // Calling setValue explicitly here instead of onChange for more control
+      // of the validation and so Formik doesn't have to infer the field name.
+      // Explicitly don't validate by setting the second param to false.
+      // Await the resulting promise to ensure the value is set before validating.
+      const maybePromise = setValue(value, false) as Promise<any> | void
+      if (maybePromise) {
+        await maybePromise
+      }
+
+      // Manually handle validation
+      if (validateOnChange) {
+        if (debouncedIdleMs) {
+          // If debouncing, set the field to touched, validate, and set
+          // isChanging to false after validating.
+          debouncedValidateAndSetIdle()
+        } else {
+          // Don't set isChanging to false if not debouncing - let the
+          // onBlur handler do that. But do validate the field.
+          validateField(name)
+        }
+      }
+    },
+    [
+      transformValueOnChange,
+      validateOnChange,
+      debouncedIdleMs,
+      validateField,
+      setValue,
+      setIsChanging,
+      debouncedValidateAndSetIdle
+    ]
+  )
+
+  /**
+   * Overrides the default Formik onBlur event in order to apply
+   * transformations and toggle isChanging appropriately.
+   */
+  const onBlur = useCallback(
+    (e: React.FocusEvent<any>) => {
+      // Apply value transformations
+      if (transformValueOnBlur) {
+        let value = getEventTargetValue(e)
+        value = transformValueOnBlur(e.target.value)
+        setValue(value, false)
+      }
+
+      // Formik's onBlur is just setTouched, but we want to clear our
+      // isChanging state once the validation finishes.
+      if (validateOnBlur) {
+        validateAndSetIdle()
+      } else {
+        setTouched(true, false)
+        setIsChanging(false)
+      }
+    },
+    [transformValueOnBlur, validateAndSetIdle]
+  )
+
+  /**
+   * Overrides the default Formik onChange event for text-like inputs.
+   *
+   * Used to apply value transformations, validation debouncing, and ensure
+   * the field name is associated properly.
+   *
+   * Since type is in scope here, avoiding the regex checks used in Formik's {@link https://github.com/jaredpalmer/formik/blob/0f960aaeeb0bdbef8312b5107cd3374884a0e62b/packages/formik/src/Formik.tsx#L639C9-L645C19 executeChange}
+   * and comparing directly for text-like input types.
+   */
+  const onChange = useCallback(
+    (e: React.ChangeEvent<any>) => {
+      if (
+        type === 'text' ||
+        type === 'email' ||
+        type === 'password' ||
+        type === 'search' ||
+        type === 'url'
+      ) {
+        onChangeText(e)
+      } else {
+        field.onChange(e)
+      }
+    },
+    [onChangeText, field.onChange, type]
+  )
+
+  // Override onChange, onBlur, and error
+  return [
+    {
+      ...field,
+      onChange,
+      onBlur
+    },
+    { ...meta, error: hasError ? error : undefined },
```

Comment:
I can think of scenarios where we may want to use different error logic than the `hasError` logic built in here (i.e. ignoring `touched` or `isChanging`), so "squashing" the old `error` here may be a bit problematic. 
The two scenarios that come to my mind right now are on the handle page we validate on mount (touched is `false), and we also validate while the user is typing (`isChanging` would be false since it hasn't been blurred yet)

Maybe we can include the original formik error still somehow?

Reply:
Ooh _excellent_ callout! I had assumed we never want to show errors when the user hasn't had a chance to type anything yet - didn't consider prefilled forms.

Will think a bit more on this. Off the cuff, I'm down to make a separate "hasError" meta field prop, but what about adding a specific check for the `validateOnMount` case? And for the other case, I think we'll want to debounce that anyway so setting `debounceIdleMs` to something nonzero should solve? And if any other scenarios come up, we can call `useField` from Formik and override the hook via the Field component's props?